### DISCUSSION TITLE
Clarification on the 2024 type system goal

### DISCUSSION THREAD

**Discussion by walkerdb** [biomejs/biome#1642] [General]
Hi all! I just read through the 2024 roadmap and I'm generally quite excited for the direction Biome is taking the ecosystem, I'm looking forward to a world where we hold our tooling to higher standards for performance and devex.

I have some concerns though around the Type System goal, some of which may just be from my own misunderstanding of the intent. Here's the main statement:

> Building a full-fledged type system such as TypeScript is a massive effort; that’s why we decided to take a different direction and start by building a subset of the type system that requires stricter typing. This approach would allow us to build some important lint rules that users have been asking for.

I would love to hear a bit more from folks on the specific goal here. From the outside this reads like Biome is planning on incrementally building a typescript competitor incompatible with current syntax. If that is the planned approach I fear that would disqualify the use of biome in the vast majority of codebases that could benefit from its adoption. Does anyone have more context on this decision and the intended direction they could share?

**Comment by Conaclos** [comment]
Hi!

The goal is not to write an incompatible type system. This could be a mistake.

My idea is to rely on the upcoming [TypeScript Isolated Declaration mode](https://github.com/microsoft/TypeScript/issues/47947) that will require more annotations from the developer and a simplified type inference system (similar to [Flow's local type inference](https://md.vern.cc/flow-type/local-type-inference-for-flow-aaa65d071347)).

We don't want to provide a type checker, just the necessary type information for implementing more powerful rules.

**Reply by walkerdb** [reply]
that makes sense to me, thank you for the clarification! I think it might be helpful to include some of these details in the roadmap post as well to keep other folks from having the same confusion
